### ISSUE TITLE
[iOS][a11y] Widgets inside OverlayPortal  don't display iOS number labels for VoiceControl

### ISSUE THREAD

**Issue by franciscojma86** [flutter/flutter#157753]
### Steps to reproduce

1. Turn Voice Control on
2. Show numbers
3. Create a widget with actionable widgets, such as buttons inside an OverlayPortal
4. Trigger the OverlayPortal

See buttons inside the OverlayPortal widget don't show numbers

b/368207734

### Expected results

Buttons should display numbers, like other buttons inside the main Widget tree

### Actual results

No iOS number labels are shown inside the OverlayPortal

### Code sample

```dart
class OverlayPortalExampleApp extends StatelessWidget {
  const OverlayPortalExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('OverlayPortal Example')),
      body: const Center(child: ClickableTooltipWidget()),
    );
  }
}

class ClickableTooltipWidget extends StatefulWidget {
  const ClickableTooltipWidget({super.key});

  @override
  State<StatefulWidget> createState() => ClickableTooltipWidgetState();
}

class ClickableTooltipWidgetState extends State<ClickableTooltipWidget> {
  final OverlayPortalController _tooltipController = OverlayPortalController();

  @override
  Widget build(BuildContext context) {
    return TextButton(
      onPressed: _tooltipController.toggle,
      child: OverlayPortal(
        controller: _tooltipController,
        overlayChildBuilder: (BuildContext context) {
          return Positioned(
            right: 50,
            bottom: 50,
            child: Material(
              color: Colors.grey,
              child: Column(
                children: [
                  TextButton(
                    onPressed: _tooltipController.toggle,
                    child: Text('Button1'),
                  ),
                  TextButton(
                    onPressed: _tooltipController.toggle,
                    child: Text('Button2'),
                  ),
                  TextButton(
                    onPressed: _tooltipController.toggle,
                    child: Text('Button3'),
                  ),
                ],
              ),
            ),
          );
        },
        child: const Text('Press to show/hide button'),
      ),
    );
  }
}
```

### Screenshots or Video

<details open>
<summary>Screenshots / Video demonstration</summary>

[Upload media here]

</details>

### Logs

<details open><summary>Logs</summary>

```console
[Paste your logs here]
```

</details>

### Flutter Doctor output

<details open><summary>Doctor output</summary>

```console
[✓] Flutter (Channel google3, on macOS 14.7 
    • Framework revision b9f62f6f9f (0 days ago), 2024-10-28T00:00:00.000
    • Engine revision 2522789c41
    • Dart version a75848f922
```

</details>

**Comment by franciscojma86** [comment]
This seems to be related https://github.com/flutter/flutter/issues/134456

**Comment by franciscojma86** [comment]
Hi @darshankawar, here's a screenshot. You have to enable Voice Control and say "show numbers". The 3 buttons should show a number, just like the "Press to show/hide" button has

![Screenshot 2024-11-11 at 3 07 04 PM](https://github.com/user-attachments/assets/9e360772-a9d8-4f93-b18a-60a7527db70e)

**Comment by franciscojma86** [comment]
b/368200688

**Comment by darshankawar** [comment]
I see, thanks for the update. I was able to replicate it using above details provided and on latest versions.

```
stable : 3.24.4
master : 3.27.0-1.0.pre.443
```

**Comment by chunhtai** [comment]
Not entirely sure what is going on, can you provide a semantics tree dump to see if the buttons are there?

**Comment by franciscojma86** [comment]
They are there! 

flutter: SemanticsNode#0
flutter:  │ Rect.fromLTRB(0.0, 0.0, 1170.0, 2532.0)
flutter:  │
flutter:  └─SemanticsNode#1
flutter:    │ Rect.fromLTRB(0.0, 0.0, 390.0, 844.0) scaled by 3.0x
flutter:    │ textDirection: ltr
flutter:    │
flutter:    └─SemanticsNode#16
flutter:      │ Rect.fromLTRB(0.0, 0.0, 390.0, 844.0)
flutter:      │ sortKey: OrdinalSortKey#1d60c(order: 0.0)
flutter:      │
flutter:      └─SemanticsNode#17
flutter:        │ Rect.fromLTRB(0.0, 0.0, 390.0, 844.0)
flutter:        │ flags: scopesRoute
flutter:        │
flutter:        ├─SemanticsNode#19
flutter:        │ │ Rect.fromLTRB(0.0, 0.0, 390.0, 103.0)
flutter:        │ │
flutter:        │ ├─SemanticsNode#20
flutter:        │ │   Rect.fromLTRB(4.0, 51.0, 52.0, 99.0)
flutter:        │ │   actions: tap
flutter:        │ │   flags: isButton, hasEnabledState, isEnabled, isFocusable
flutter:        │ │   tooltip: "Back"
flutter:        │ │   textDirection: ltr
flutter:        │ │
flutter:        │ └─SemanticsNode#21
flutter:        │     Rect.fromLTRB(72.0, 57.5, 327.7, 92.5)
flutter:        │     flags: isHeader
flutter:        │     label: "OverlayPortal Example"
flutter:        │     textDirection: ltr
flutter:        │
flutter:        └─SemanticsNode#18
flutter:          │ Rect.fromLTRB(74.1, 449.5, 315.9, 497.5)
flutter:          │ actions: tap
flutter:          │ flags: isButton, hasEnabledState, isEnabled, isFocusable
flutter:          │ label: "Press to show/hide button"
flutter:          │ textDirection: ltr
flutter:          │
flutter:          ├─SemanticsNode#22
flutter:          │   Rect.fromLTRB(177.8, 200.5, 264.3, 248.5)
flutter:          │   actions: tap
flutter:          │   flags: isButton, hasEnabledState, isEnabled, isFocusable
flutter:          │   label: "Button1"
flutter:          │   textDirection: ltr
flutter:          │
flutter:          ├─SemanticsNode#23
flutter:          │   Rect.fromLTRB(176.4, 248.5, 265.7, 296.5)
flutter:          │   actions: tap
flutter:          │   flags: isButton, hasEnabledState, isEnabled, isFocusable
flutter:          │   label: "Button2"
flutter:          │   textDirection: ltr
flutter:          │
flutter:          └─SemanticsNode#24
flutter:              Rect.fromLTRB(176.2, 296.5, 265.9, 344.5)
flutter:              actions: tap
flutter:              flags: isButton, hasEnabledState, isEnabled, isFocusable
flutter:              label: "Button3"
flutter:              textDirection: ltr

**Comment by franciscojma86** [comment]
b/365916888

**Comment by loic-sharma** [comment]
@LouiseHsu and I did some digging here. It appears that Voice Control does not show a number for an accessibility element that is outside of its parents accessibility bounds, even if the accessibility element can be interacted by the user. Unfortunately, Flutter's `Overlay` creates such accessibility trees.

<details>
<summary>Minimal Dart app...</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const OverlayPortalExampleApp());
}

class OverlayPortalExampleApp extends StatelessWidget {
  const OverlayPortalExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: const Center(child: ClickableTooltipWidget()),
      ),
    );
  }
}

class ClickableTooltipWidget extends StatefulWidget {
  const ClickableTooltipWidget({super.key});

  @override
  State<StatefulWidget> createState() => ClickableTooltipWidgetState();
}

class ClickableTooltipWidgetState extends State<ClickableTooltipWidget> {
  final OverlayPortalController _tooltipController = OverlayPortalController();

  @override
  Widget build(BuildContext context) {
    return TextButton(
      onPressed: _tooltipController.toggle,
      child: OverlayPortal(
        controller: _tooltipController,
        overlayChildBuilder: (BuildContext context) {
          return Positioned(
            right: 50,
            bottom: 50,
            child: TextButton(
              onPressed: _tooltipController.toggle,
              child: Text('Button1'),
            ),
          );
        },
        child: const Text('Press to show/hide button'),
      ),
    );
  }
}
```

</details>

<details>
<summary>Flutter semantics tree...</summary>

```
SemanticsNode#0
 │ Rect.fromLTRB(0.0, 0.0, 1179.0, 2556.0)
 │
 └─SemanticsNode#1
   │ Rect.fromLTRB(0.0, 0.0, 393.0, 852.0) scaled by 3.0x
   │ textDirection: ltr
   │
   └─SemanticsNode#2
     │ Rect.fromLTRB(0.0, 0.0, 393.0, 852.0)
     │ sortKey: OrdinalSortKey#64743(order: 0.0)
     │
     └─SemanticsNode#3
       │ Rect.fromLTRB(0.0, 0.0, 393.0, 852.0)
       │ flags: scopesRoute
       │
       └─SemanticsNode#4
         │ Rect.fromLTRB(95.3, 402.0, 297.7, 450.0)
         │ actions: tap
         │ flags: isButton, hasEnabledState, isEnabled, isFocusable
         │ label: "Press to show/hide button"
         │ textDirection: ltr
         │
         └─SemanticsNode#5
             Rect.fromLTRB(171.5, 352.0, 247.7, 400.0)
             actions: tap
             flags: isButton, hasEnabledState, isEnabled, isFocusable
             label: "Button1"
             textDirection: ltr
```

`Button1` is the button in the overlay.

</details>

<details>
<summary>Repro steps...</summary>

To repro:

1. Run the app
2. Press `Press to show/hide button`. The `Button1` button should be revealed.
3. Turn on Voice Control. Say "Show numbers". A number appears above `Press to/show hide button` but not above `Button1`.

</details>

The iOS accessibility tree looks like:

```
SemanticsObjectContainer
 └─Press to show/hide button (button)
 └─Button1 (button)
```

The parent `SemanticsObjectContainer` has the accessibility bounds of the `Press to show/hide button` button. As a result, `Button1` button is outside of its parent's accessibility bounds. 

@LouiseHsu noticed that if we change the overlay child's `Positioned` such that `Button1` is close to but not on top of `Press to show/hide button`, the Voice Control number works as expected:

![Image](https://github.com/user-attachments/assets/dde3a0ad-f8f2-4d15-bc5d-d88e08fc50b3)

We will need to generate a different accessibility tree for the `Overlay` scenario. cc @chunhtai

**Comment by chunhtai** [comment]
yes... the overlayportal is a special case where the renderobject tree will be different from semantics tree, and they are intentionly done this way so that one can create a pop up outside its bound but still share the same BuildContext.

basically in element tree you have this
```
Overlay
 |
Textbutton
 |
OverlayPortal
```
the overlay will however attach itself to overlay in rendering tree
so it will look like this in rendering tree
```
RenderOverlay
|                          \
RenderButton       _RenderLayoutSurrogateProxyBox
```

but then because logically overlayportal should be under the button, so the semantics node of  _RenderLayoutSurrogateProxyBox will attach its self back to the semantics node of RenderButton
```
SemanticsNode(Overlay)
 |
SemanticsNode(Textbutton)
 |
SemanticsNode(OverlayPortal)
```

This means the overlayportal semantics node may be bigger than parent is WAI. we can't expand the parent to compensate it, because that will cause the text button to have incorrect bound.

Since this is really a issue for iOS only, can the SemanticsObjectContainer to compensate this instead?
